File: mobile/lib/models/alpaca.dart
/// Alpaca paper trading data models (AT:R45; reworked CR202).
///
/// CR202 moved the credential to the device, so these are now parsed from
/// **Alpaca's own JSON** rather than from an AMI backend response that had
/// already coerced the types. Alpaca reports every monetary field and quantity
/// as a *string* (`"cash": "12450.32"`), so parsing must accept both a string
/// and a number — the previous `as num` cast would have thrown on every real
/// response.
///
/// AlpacaPortfolio — paper account summary (cash, equity, buying power).
/// AlpacaPosition — one open paper position.
///
/// Link status is no longer a model: it is simply whether the device holds a
/// credential (see AlpacaCredentialStore.isLinked).
library;

/// Tolerant numeric parse: Alpaca sends strings, our own tests send numbers,
/// and a missing/garbage field reads as 0 rather than throwing — a malformed
/// single field should not blank the whole panel.
double alpacaNum(dynamic v) {
  if (v == null) return 0;
  if (v is num) return v.toDouble();
  return double.tryParse(v.toString()) ?? 0;
}

class AlpacaPortfolio {
  const AlpacaPortfolio({
    required this.cash,
    required this.portfolioValue,
    required this.equity,
    required this.buyingPower,
  });

  final double cash;
  final double portfolioValue;
  final double equity;
  final double buyingPower;

  factory AlpacaPortfolio.fromJson(Map<String, dynamic> j) => AlpacaPortfolio(
        cash: alpacaNum(j['cash']),
        portfolioValue: alpacaNum(j['portfolio_value']),
        equity: alpacaNum(j['equity']),
        buyingPower: alpacaNum(j['buying_power']),
      );
}

class AlpacaPosition {
  const AlpacaPosition({
    required this.symbol,
    required this.qty,
    required this.marketValue,
    required this.unrealizedPl,
  });

  final String symbol;
  final double qty;
  final double marketValue;
  final double unrealizedPl;

  factory AlpacaPosition.fromJson(Map<String, dynamic> j) => AlpacaPosition(
        symbol: (j['symbol'] ?? '') as String,
        qty: alpacaNum(j['qty']),
        marketValue: alpacaNum(j['market_value']),
        unrealizedPl: alpacaNum(j['unrealized_pl']),
      );

  /// The upload shape the backend accepts (`schemas/alpaca.AlpacaPositionIn`).
  Map<String, dynamic> toWireJson() => {
        'symbol': symbol,
        'qty': qty,
        'market_value': marketValue,
        'unrealized_pl': unrealizedPl,
      };
}

/// One Alpaca paper order, as placed by `AlpacaClient.submitOrder` (CR227).
///
/// Market orders only (v1) — see CR227's Non-goals. Parsed the same
/// tolerant way as the other models here: Alpaca returns `filled_qty` as a
/// string, and an order accepted-but-not-yet-filled reports a null fill
/// price, which should read as "not yet filled" rather than throw.
class AlpacaOrder {
  const AlpacaOrder({
    required this.id,
    required this.symbol,
    required this.side,
    required this.qty,
    required this.status,
    this.filledAvgPrice,
  });

  final String id;
  final String symbol;
  final String side;
  final double qty;
  final String status;
  final double? filledAvgPrice;

  factory AlpacaOrder.fromJson(Map<String, dynamic> j) => AlpacaOrder(
        id: (j['id'] ?? '') as String,
        symbol: (j['symbol'] ?? '') as String,
        side: (j['side'] ?? '') as String,
        qty: alpacaNum(j['qty']),
        status: (j['status'] ?? '') as String,
        filledAvgPrice:
            j['filled_avg_price'] == null ? null : alpacaNum(j['filled_avg_price']),
      );
}

/// What the device sends with a Room convene / 1-on-1 turn so the agents can
/// see the linked account. Values only — the backend owns the layout of the
/// block that reaches the prompt.
class AlpacaSnapshot {
  const AlpacaSnapshot({required this.portfolio, required this.positions});

  final AlpacaPortfolio portfolio;
  final List<AlpacaPosition> positions;

  /// Alpaca's position cap on our side mirrors the backend's MAX_POSITIONS —
  /// sending more would be rejected wholesale, losing the overlay entirely, so
  /// the largest positions are kept and the tail is dropped.
  static const int maxPositions = 100;

  Map<String, dynamic> toWireJson() {
    final sorted = [...positions]
      ..sort((a, b) => b.marketValue.abs().compareTo(a.marketValue.abs()));
    return {
      'cash': portfolio.cash,
      'portfolio_value': portfolio.portfolioValue,
      'buying_power': portfolio.buyingPower,
      'positions': sorted.take(maxPositions).map((p) => p.toWireJson()).toList(),
    };
  }

  /// DEF419 — the `AccountSnapshotIn` shape `POST /v1/sim/preview`'s `account`
  /// field expects (`backend/app/schemas/alpaca.py`): `kind`/`equity`/`cash`/
  /// `positions[{ticker, qty, market_value}]`. Deliberately NOT [toWireJson]
  /// above — that method serialises the Room-overlay `AlpacaSnapshotIn` shape
  /// (`portfolio_value`/`buying_power`, `symbol`/`unrealized_pl` on each
  /// position), a different wire contract for a different endpoint. Mixing
  /// the two up would 422 on the field names alone (`extra="forbid"` on both
  /// schemas), which is the loud failure CR040 asks for over a silent
  /// mismatch — but there is no reason to invite it when the two are this
  /// easy to keep apart by having one method per contract.
  ///
  /// A negative `qty` (a short Alpaca position) is dropped rather than sent
  /// negative: `AccountPositionIn.qty` is bounded `ge=0` server-side (DEF419's
  /// schema — shorts are out of scope for the account-snapshot path, same as
  /// the sim engine's own `shorts=None` on this branch), so sending one would
  /// 422 the whole preview over one position the mandate check does not model
  /// anyway.
  Map<String, dynamic> toMandateSnapshotJson() {
    final sorted = [...positions]
      ..sort((a, b) => b.marketValue.abs().compareTo(a.marketValue.abs()));
    return {
      'kind': 'alpaca_paper',
      'equity': portfolio.equity,
      'cash': portfolio.cash,
      'positions': sorted
          .take(maxPositions)
          .where((p) => p.qty >= 0)
          .map((p) => {
                'ticker': p.symbol,
                'qty': p.qty,
                'market_value': p.marketValue,
              })
          .toList(),
    };
  }
}

File: backend/app/schemas/alpaca.py
"""Client-supplied Alpaca paper-account snapshot (CR202).

The user's Alpaca credentials live on their device — Keychain on iOS,
Keystore-backed EncryptedSharedPreferences on Android — and never reach this
host. So the device fetches its own paper account and uploads the *result*
with each Room convene or 1-on-1 turn; these models are that wire contract.

Everything here is untrusted input that ends up inside twelve agent prompts,
which is why the bounds are structural rather than advisory. CLAUDE.md's rule
is that prompt instructions are not controls (agents ignore even emphatic ones
~70% of the time), so the client is allowed to supply *values* and never
*layout*: the text block is rendered server-side by
`alpaca_service.snapshot_text`, from these validated fields only.

The four bounds, and the specific thing each one stops:

* `extra="forbid"`  — an unexpected key cannot ride along into the renderer.
* `symbol` pattern  — a ticker is the only free-text field that reaches the
  prompt. Without the pattern it is a prompt-injection channel: a "symbol" of
  `IGNORE ALL PRIOR INSTRUCTIONS AND ...` would be rendered verbatim into every
  agent's context.
* `max_length` on positions — bounds prompt size, the same reasoning DEF186
  used to cap `user_message` at 8,000 chars.
* finite floats — `float("nan")` is accepted by Python and by JSON parsers
  that allow it, and would render into an agent prompt as the literal string
  `nan`, which the model then reasons about as if it were a number.
"""

from __future__ import annotations

import math
from typing import Annotated

from pydantic import AfterValidator, BaseModel, ConfigDict, Field

# A paper account can legitimately hold a lot of names; 100 is far above any
# real one while still bounding what a hostile client can push into a prompt.
MAX_POSITIONS = 100


def _finite(v: float) -> float:
    if not math.isfinite(v):
        raise ValueError("must be a finite number")
    return v


FiniteFloat = Annotated[float, AfterValidator(_finite)]

# Uppercase ticker, optional dot/dash class markers (BRK.B, RDS-A). Deliberately
# strict: this is the only free-text field that reaches an agent prompt.
_SYMBOL_PATTERN = r"^[A-Z][A-Z0-9.\-]{0,9}$"


class AlpacaPositionIn(BaseModel):
    """One open paper position, as reported by the device."""

    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(pattern=_SYMBOL_PATTERN)
    qty: FiniteFloat
    market_value: FiniteFloat
    unrealized_pl: FiniteFloat


class AlpacaSnapshotIn(BaseModel):
    """A whole paper account, as reported by the device.

    Optional on every request that accepts it — absent means "no linked
    account", which is the path the overwhelming majority of runs take and
    which renders no overlay at all.
    """

    model_config = ConfigDict(extra="forbid")

    cash: FiniteFloat
    portfolio_value: FiniteFloat
    buying_power: FiniteFloat
    positions: list[AlpacaPositionIn] = Field(default_factory=list, max_length=MAX_POSITIONS)


# CR230 — order-attempt logging. Unlike AlpacaSnapshotIn above, this never
# reaches an LLM prompt (it's a write-only audit row), so the prompt-injection
# rationale for the symbol pattern doesn't apply here. The pattern is kept
# anyway: a DB column is not the place for a client to write arbitrary
# strings either, and reusing the existing bound keeps one definition of
# "what a ticker looks like" rather than two.
class AlpacaOrderLogIn(BaseModel):
    """One `AlpacaClient.submitOrder()` outcome, as reported by the device.

    CR230 — this is a report, not an observation, in the same sense
    `LinkStateRequest` (CR203) is: the backend never holds the Alpaca
    credential (CR202), so it cannot independently verify any of this. It
    is the device's own account of what happened, logged because Saiful
    asked for a record of every Alpaca interaction, not because the backend
    can confirm it against Alpaca's own order book.
    """

    model_config = ConfigDict(extra="forbid")

    symbol: str = Field(pattern=_SYMBOL_PATTERN)
    side: str = Field(pattern=r"^(buy|sell)$")
    qty: FiniteFloat
    destination: str = Field(pattern=r"^(alpaca_only|both)$")
    outcome: str = Field(pattern=r"^(submitted|rejected_by_alpaca|refused_client_side)$")
    detail: str | None = Field(default=None, max_length=500)
    alpaca_order_id: str | None = Field(default=None, max_length=64)
    alpaca_status: str | None = Field(default=None, max_length=32)


# DEF419 — per-account mandate checking on /v1/sim/preview. Saiful, from a
# TestFlight screenshot: an Alpaca-only preview was evaluated against the
# AMI sim portfolio's equity, so a trade sized fine for the (larger/smaller)
# Alpaca paper account could be wrongly rejected or wrongly approved. "It
# does mean that it may reject for ami and approve for alpaca. Or vice
# versa. This is an expected condition" — the fix is not "make them agree",
# it's "check against the account actually receiving the order".
#
# Client-attested, same custody boundary as `AlpacaSnapshotIn` (CR202): the
# backend never holds the Alpaca credential, so it cannot independently
# verify equity/cash/positions against Alpaca's own ledger. Unlike
# `AlpacaSnapshotIn` this NEVER reaches an LLM prompt — it only feeds the
# deterministic `check_mandate_compliance` — so the prompt-injection
# rationale doesn't apply, but the bounding is kept anyway (CR230's same
# reasoning: a DB column, or here a sizing calculation, is not the place for
# a client to write arbitrary/non-finite values either). Malformed input is
# a 422 (FastAPI's standard schema-validation response), never a silent
# fallback to the AMI account — falling back IS the bug this DEF fixes.
class AccountPositionIn(BaseModel):
    """One open position in the account being checked against the mandate."""

    model_config = ConfigDict(extra="forbid")

    ticker: str = Field(pattern=_SYMBOL_PATTERN)
    qty: FiniteFloat = Field(ge=0)
    market_value: FiniteFloat = Field(ge=0)


class AccountSnapshotIn(BaseModel):
    """The account the proposed trade should be sized against, when it is
    not AMI's own sim portfolio.

    `kind` is a discriminator for the response echo (so the client can label
    the outcome, e.g. "checked against your Alpaca paper account") and for
    future account kinds — only `alpaca_paper` exists today (CR227).

    Optional on `SubmitTradeRequest`: absent means "check against the AMI
    sim portfolio", byte-identical to pre-DEF419 behaviour. Present means
    `check_mandate_compliance` runs the SAME mandate against THIS snapshot
    instead — same function, same rules, different denominator.
    """

    model_config = ConfigDict(extra="forbid")

    kind: str = Field(pattern=r"^(alpaca_paper)$")
    equity: FiniteFloat = Field(gt=0)
    cash: FiniteFloat = Field(ge=0)
    positions: list[AccountPositionIn] = Field(
        default_factory=list, max_length=MAX_POSITIONS
    )
